d2ir/import.go

```diff
@@ -0,0 +1,98 @@
+package d2ir
+
+import (
+	"bufio"
+	"path"
+
+	"oss.terrastruct.com/d2/d2ast"
+	"oss.terrastruct.com/d2/d2parser"
+)
+
+func (c *compiler) pushImportStack(imp *d2ast.Import) bool {
+	if len(imp.Path) == 0 {
+		c.errorf(imp, "imports must specify a path to import")
+		return false
+	}
+
+	newPath := imp.Path[0].Unbox().ScalarString()
+	for _, p := range c.importStack {
+		if newPath == p {
+			c.errorf(imp, "detected cyclic import of %q", newPath)
+			return false
+		}
+	}
+
+	c.importStack = append(c.importStack, newPath)
+	return true
+}
+
+func (c *compiler) popImportStack() {
+	c.importStack = c.importStack[:len(c.importStack)-1]
+}
+
+// Returns either *Map or *Field.
+func (c *compiler) _import(imp *d2ast.Import) (Node, bool) {
+	ir, ok := c.__import(imp)
+	if !ok {
+		return nil, false
+	}
+	if len(imp.IDA()) > 0 {
+		f := ir.GetField(imp.IDA()...)
+		if f == nil {
+			c.errorf(imp, "import key %q doesn't exist inside import", imp.IDA())
+			return nil, false
+		}
+		return f, true
+	}
+	return ir, true
+}
+
+func (c *compiler) __import(imp *d2ast.Import) (*Map, bool) {
+	impPath := imp.Path[0].Unbox().ScalarString()
+	if path.IsAbs(impPath) {
+		c.errorf(imp, "import paths must be relative")
+		return nil, false
+	}
```

Comment:
why?

Reply:
We can change it later if we want to but I don't see why anyone would want an actual absolute import path?

I was thinking maybe we have stuff like typescript's import aliases later so you can import without having to specify all the relative paths. Or maybe define absolute import paths to be from the git root directory.

Reply:
i think variables work well enough. 

```d2
$ROOT: ../dir/something

x: @$ROOT/x
```

Reply:
Well the parser doesn't support that now but yea that could be an option.